### PR TITLE
RavenDB-24327 Persist freed pages in journal files  to skip their checksum validation after restart.

### PR DESCRIPTION
**Pull request by arekpalinski**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-24327/Invalid-checksum-error-after-process-termination

### Additional description

Added support for tracking freed pages in transactions and persisting this information in the journal. This ensures proper recovery handling by skipping checksum checks on freed pages, in particular those that were part of a sparse region.

### Type of change

- [x] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [ ] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [ ] Non breaking change
- [x] Ensured. Please explain how has it been implemented?
   - Journal entries that include free pages set a new flag in a transaction header
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing
- [ ] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

### CODE REVIEW

**Comment by arekpalinski** [comment]
@ravendb-bot test this in debug please

**Comment by arekpalinski** [comment]
Please hold on with the review

**Comment by arekpalinski** [comment]
Ready for the review

File: src/Voron/Impl/Journal/WriteAheadJournal.cs [hunk 1]
```diff
@@ -2084,8 +2057,31 @@ private Pal.journal_entry PrepareToWriteToJournal(LowLevelTransaction tx, ref Pa
...
+                
+                foreach (var page in txPages)
+                {
+                    if (freedPages.Remove(page.PageNumberInDataFile) && page.NumberOfPages > 1)
+                    {
+                        for (int i = 1; i < page.NumberOfPages; i++)
+                        {
+                            freedPages.Remove(page.PageNumberInDataFile + i);
+                        }
+                    }
```
Comment (ayende) [inline src/Voron/Impl/Journal/WriteAheadJournal.cs]: Shouldn't this be something that we are handling during the `ModifyPage`, instead?

File: src/Voron/Impl/Journal/TransactionHeader.cs [hunk 2]
```diff
@@ -92,4 +92,11 @@ public override string ToString()
                    $" UncompressedSize: {UncompressedSize}, TimeStamp: {timestamp}";
         }
     }
-}
\ No newline at end of file
+    
+    [StructLayout(LayoutKind.Explicit, Pack = 1)]
```
Comment (redknightlois) [inline src/Voron/Impl/Journal/TransactionHeader.cs]: `LayoutKind.Explicit` makes `Pack` irrelevant. In this case, `Pack` is a no-op can be safely removed.

File: src/Voron/Impl/Journal/WriteAheadJournal.cs (hunk 1 above)
Comment (redknightlois) [reply src/Voron/Impl/Journal/WriteAheadJournal.cs]: Pushing this into the transaction would require us to keep the state in sync during a high traffic operation (`AllocatePageImpl`) while we can do it here with `O(#tx pages + size of freed set)`. I don't know, from the performance point of view it could make sense to push it earlier since we have `O( 1 )` on average by accessing the `HashSet` anyways but definitely something that should be shown with performance numbers in hand. I see potential cases where it would be interesting to see like on batch deletes.

File: src/Voron/Impl/Journal/WriteAheadJournal.cs [hunk 3]
```diff
@@ -2084,8 +2057,31 @@ private Pal.journal_entry PrepareToWriteToJournal(LowLevelTransaction tx, ref Pa
...
 
+            var freedPages = tx.GetFreedPages();
+
+            if (freedPages is { Count: > 0 })
+            {
+                // we need to remove the freed pages that were reused and modified later on during the transaction
+                
+                foreach (var page in txPages)
+                {
+                    if (freedPages.Remove(page.PageNumberInDataFile) && page.NumberOfPages > 1)
```
Comment (redknightlois) [inline src/Voron/Impl/Journal/WriteAheadJournal.cs]: Invert this, we want to use the short-circuit to avoid the HashSet operation when skipping the loop completely is possible via `page.NumberOfPages`

File: src/Voron/Impl/Journal/WriteAheadJournal.cs (hunk 1 above)
Comment (ayende) [reply src/Voron/Impl/Journal/WriteAheadJournal.cs]: The problem is that it is entirely feasible to have a transaction in which we:

* modify the page
* free it
* allocate it again
* free it again

Etc...

That is a problem for this scenario, because you cannot assume that modified page / free pages are exclusive.

File: src/Voron/Impl/Journal/TransactionHeader.cs (hunk 2 above)
Comment (ayende) [reply src/Voron/Impl/Journal/TransactionHeader.cs]: Note, we should probably `pack` this using `pfor` or the like.

File: src/Voron/Impl/Journal/WriteAheadJournal.cs [hunk 4]
```diff
@@ -2084,8 +2057,31 @@ private Pal.journal_entry PrepareToWriteToJournal(LowLevelTransaction tx, ref Pa
...
+                        }
+                    }
+                }
+            }
+            
+            var numberOfFreedPages = freedPages?.Count ?? 0;
+            
+            var freedPagesOverhead = numberOfFreedPages * sizeof(TransactionFreePage);
+            var freedPagesOverheadInPages = checked(freedPagesOverhead / Constants.Storage.PageSize + (freedPagesOverhead % Constants.Storage.PageSize == 0 ? 0 : 1));
+
```
Comment (ayende) [inline src/Voron/Impl/Journal/WriteAheadJournal.cs]: freedPagesOverhead = encoder.Encode(l, array.Length);

Comment (ayende) [reply src/Voron/Impl/Journal/WriteAheadJournal.cs]: (int count, int sizeUsed) = encoder.Write(buffer, Container.MaxSizeInsideContainerPage);

File: src/Voron/Impl/Journal/WriteAheadJournal.cs (hunk 1 above)
Comment (redknightlois) [reply src/Voron/Impl/Journal/WriteAheadJournal.cs]: We need a test for that. That's a very interesting edge case I was not thinking about.

Comment (arekpalinski) [reply src/Voron/Impl/Journal/WriteAheadJournal.cs]: After direct discussion with @ayende we agreed to handle it this way. It's way simpler approach than syncing transaction pages and free ones during transaction. Also the same approach we had in versions before 8.0.

The code itself got fixed and the test case was added.

File: src/Voron/Impl/Journal/TransactionHeader.cs (hunk 2 above)
Comment (arekpalinski) [reply src/Voron/Impl/Journal/TransactionHeader.cs]: Removed `Pack`. The code got changed to use PFor encoding.

File: src/Voron/Impl/Journal/WriteAheadJournal.cs (hunk 3 above)
Comment (arekpalinski) [reply src/Voron/Impl/Journal/WriteAheadJournal.cs]: This got removed

**Comment by arekpalinski** [comment]
@ravendb-bot test this in debug please

**Review by redknightlois — APPROVED** [review]
I like that we just needed to impact `FreePage` with an O( 1 ) on average operation. LGTM

File: src/Voron/Impl/Journal/JournalReader.cs [hunk 5]
```diff
@@ -237,11 +243,76 @@ public bool ReadOneTransactionToDataFile(ref Pager.State dataPagerState, ref Pag
...
 
+        internal static List<long> ReadEncodedFreePage(byte* src, ByteStringContext allocator)
+        {
+            var readPtr = src;
+
+            var headerPtr = readPtr;
+
+            var header = *(FreePagesHeader*)headerPtr;
+            var sectionHeaders = (EncodedFreePagesSection*)(headerPtr + sizeof(FreePagesHeader) + header.EncodedSectionsSize);
+
```
Comment (ayende) [inline src/Voron/Impl/Journal/JournalReader.cs:279]: Let's add some checks here that we aren't trying to read past the transaction size.

File: src/Voron/Impl/Journal/JournalReader.cs [hunk 6]
```diff
@@ -237,11 +243,76 @@ public bool ReadOneTransactionToDataFile(ref Pager.State dataPagerState, ref Pag
...
+                
+            long* freePages = (long*)buffer.Ptr;
+
+            var totalCount = 0;
+            
+            for (int i = 0; i < header.EncodedSectionsCount; i++)
+            {
+                var sectionSize = sectionHeaders[i].Size;
+                
+                using var reader = new FastPForBufferedReader(allocator);
```
Comment (ayende) [inline src/Voron/Impl/Journal/JournalReader.cs:304]: Does this free all the memory?
Do we need to *reset* the allocator between journals?

File: src/Voron/Impl/Journal/WriteAheadJournal.cs [hunk 7]
```diff
@@ -353,7 +354,7 @@ public bool RecoverDatabase(TransactionHeader* txHeader, out long lastJournalNum
 
                     List<TransactionHeader> transactionHeaders;
                     var journalReader = new JournalReader(_env, journalNumber, journalPager, journalPagerState, dataPager, recoveryPager, modifiedPages, logInfo, currentFileHeader,
-                        transactionHeader);
+                        transactionHeader, allocator);
```
Comment (ayende) [inline src/Voron/Impl/Journal/WriteAheadJournal.cs:359]: can we reset here to avoid leaks?

File: src/Voron/Impl/Journal/WriteAheadJournal.cs [hunk 8]
```diff
@@ -2268,6 +2273,9 @@ private Pal.journal_entry PrepareToWriteToJournal(LowLevelTransaction tx, ref Pa
             txHeader.PageCount = numberOfPages;
             txHeader.JournalId = _headerAccessor.JournalId;
             
+            if (freedPages is { Count: > 0 })
+                txHeader.Flags |= TransactionPersistenceModeFlags.HasFreePages;
```
Comment (ayende) [inline src/Voron/Impl/Journal/WriteAheadJournal.cs]: Can we move this to line `2180`, where we also have this?

File: src/Voron/Impl/Journal/WriteAheadJournal.cs [hunk 9]
```diff
@@ -2491,5 +2499,66 @@ internal sealed class TestingStuff
         }
 
         private void RejectCommitsToMerge() => SharedJournalState.SetCancel();
+        
+        internal static long EncodeFreePages(HashSet<long> freedPages, byte* dst, ByteStringContext allocator)
+        {
+            var pages = freedPages.ToArray();
```
Comment (ayende) [inline src/Voron/Impl/Journal/WriteAheadJournal.cs]: we already have an allocator, let's use that instead of allocating managed array.
I would like to avoid this allocation each time

File: src/Voron/Impl/Journal/JournalReader.cs (hunk 5 above)
Comment (arekpalinski) [reply src/Voron/Impl/Journal/JournalReader.cs:279]: Added

File: src/Voron/Impl/Journal/JournalReader.cs (hunk 6 above)
Comment (arekpalinski) [reply src/Voron/Impl/Journal/JournalReader.cs:304]: Yes, it does. Anyway the allocator is reset now between journals

File: src/Voron/Impl/Journal/WriteAheadJournal.cs (hunk 7 above)
Comment (arekpalinski) [reply src/Voron/Impl/Journal/WriteAheadJournal.cs:359]: Done

File: src/Voron/Impl/Journal/WriteAheadJournal.cs (hunk 8 above)
Comment (arekpalinski) [reply src/Voron/Impl/Journal/WriteAheadJournal.cs]: Done

File: src/Voron/Impl/Journal/WriteAheadJournal.cs (hunk 9 above)
Comment (arekpalinski) [reply src/Voron/Impl/Journal/WriteAheadJournal.cs]: Done

File: src/Voron/Impl/Journal/JournalReader.cs [hunk 10]
```diff
@@ -245,7 +245,20 @@ public bool ReadOneTransactionToDataFile(ref Pager.State dataPagerState, ref Pag
...
+                var freePagesSpaceToRead = sizeof(FreePagesHeader) + header.EncodedSectionsSize + header.EncodedSectionsCount * sizeof(EncodedFreePagesSection);
+
+                if (totalRead + freePagesSpaceToRead > current->UncompressedSize)
+                {
+                    throw new InvalidDataException(
+                        $"Attempted to read free pages from position {totalRead} to {totalRead + freePagesSpaceToRead} (size: {freePagesSpaceToRead} bytes) from transaction data while " +
+                        $"the transaction uncompressed size is {current->UncompressedSize} bytes");
+                }
+
+                var freePages = ReadEncodedFreePage(freePagesPtr, _allocator);
```
Comment (ayende) [inline src/Voron/Impl/Journal/JournalReader.cs]: I think we need to pass `UncompressedSize` here as well.
If we have a bad value in one of the segments size, we need to error there as well

Comment (arekpalinski) [reply src/Voron/Impl/Journal/JournalReader.cs]: Validation of section sizes added. I put entire validation code inside `ReadEncodedFreePage`

File: src/Voron/Impl/Journal/WriteAheadJournal.cs [hunk 11]
```diff
@@ -2491,5 +2501,72 @@ internal sealed class TestingStuff
         }
 
         private void RejectCommitsToMerge() => SharedJournalState.SetCancel();
+
+        internal static long EncodeFreePages(HashSet<long> freedPages, byte* dst, ByteStringContext allocator)
+        {
```
Comment (redknightlois) [inline src/Voron/Impl/Journal/WriteAheadJournal.cs:2529]: Can we add somewhere a bit of information about the format for the journal frame? I am thinking in terms of ourselves 4 years from now.

Comment (arekpalinski) [reply src/Voron/Impl/Journal/WriteAheadJournal.cs:2529]: I will add a comment showing the format of a journal entry

File: src/Voron/Impl/Journal/WriteAheadJournal.cs [hunk 12]
```diff
@@ -2491,5 +2521,72 @@ internal sealed class TestingStuff
...
+
+            var encodedSize = 0;
+
+            var sectionSizes = new List<EncodedFreePagesSection>(1);
+
+            while (remainingCount > 0)
+            {
+                encoder.Encode(pages + processedCount, remainingCount);
+
+                var (count, size) = encoder.Write(output, maxEncodingSectionSize);
```
Comment (ayende) [inline src/Voron/Impl/Journal/WriteAheadJournal.cs]: One last thing - here, can we verify that we have sufficient space?
I _think_ we must have, because we allocated twice as much, but it should be checked.
Also, we may want to error if `count` is 0, etc.

Comment (arekpalinski) [reply src/Voron/Impl/Journal/WriteAheadJournal.cs]: Done. Please check https://github.com/ravendb/ravendb/pull/21900/commits/b74cf814714454aaec381df25bac63849d35619d

**Review by ayende — APPROVED** [review]
